src/doorman/TelnetProtocol.hpp

```diff
@@ -0,0 +1,46 @@
+#pragma once
+
+#include "Byte.hpp"
+
+#include <gsl/span>
+
+#include <vector>
+
+class TelnetProtocol {
+public:
+    struct Handler {
+        virtual ~Handler() = default;
+        virtual void send_bytes(gsl::span<const byte> data) = 0;
+        virtual void on_line(std::string_view line) = 0;
+        virtual void on_terminal_size(int width, int height) = 0;
+        virtual void on_terminal_type(std::string_view type, bool ansi_supported) = 0;
+    };
+
+private:
+    Handler &handler_;
+    std::vector<byte> buffer_;
+    bool got_term_ = false;
+    bool echoing_ = true;
+    int width_ = 80;
+    int height_ = 24;
+    bool ansi_ = false;
+    char previous_separator_ = 0;
+
+    void send_com(byte a, byte b);
+    void send_opt(byte a);
+
+    size_t on_subcommand(gsl::span<const byte> command_sequence);
```

Comment:
A comment about what the return value means would be good. I guess it's the length of the command in bytes, but explicit is nice. (same for on_command)

Reply:
Done